### ISSUE TITLE
[RELEASE BLOCKER][1.24] Golang 1.18 issues

### ISSUE THREAD

**Issue by justaugustus** [kubernetes/kubernetes#108357]
Upstream/downstream tracking issues:

- [ ] Release engineering tracking for go1.18: https://github.com/kubernetes/release/issues/2307
- [ ] ...
- [ ] ...
- [ ] ...

cc: @kubernetes/release-engineering @kubernetes/release-team @dims @liggitt 

---

From https://github.com/kubernetes/kubernetes/pull/107105#issuecomment-1049557807:

> It seems that this PR has visibly increase memory usage of at least kube-apiserver (10% on 10th-percentile, even more on higher percentiles): https://perf-dash.k8s.io/#/?jobname=gce-5000Nodes&metriccategoryname=E2E&metricname=LoadResources&PodName=kube-apiserver-gce-scale-cluster-master%2Fkube-apiserver&Resource=memory
> 
> We should understand to what extent this is expected and can we somehow mitigate this.

> @wojtek-t can we file an issue to golang so they're aware as we investigate? do we have any pprof info to help narrow down _where_ the memory increase is?

> We're following up with golang team about that. @tosi3k is looking into it and started the conversation with golang folks.

> @liggitt @wojtek-t do we want to open a blocker issue in k/k for tracking the problem above?

**Comment by wojtek-t** [comment]
@kubernetes/sig-scalability @mborsz @tosi3k 

Just to copy the current hypothesis from our discussion with golang folks

```
Yes, this is almost certainly because of the change to GC scheduling in Go 1.18.

The most likely effect of that that would cause a significant increase in memory is if you have a lot of stack memory relative to heap memory. 
```

This is under further investigation.

**Comment by wojtek-t** [comment]
We believe we already understand the cause of apiserver`s increased memory usage and this is related to golang GC changes.

Citing the summary from golang folks:

```
the middle line [memory usage graph from wojtekt`s original comment] jumps from ~33000 MB to ~40000 MB (there's no key, so I picked that line arbitrarily). From the stack/heap ratio graph [attached below], we see a ratio between 15% and 30%. Taking the middle of that range, we would expect a previous 33000 MB footprint to grow to 33000 * 1.225 = 40425 MB, which is exactly what we see. So this is working exactly as intended from Go's side.

Since this issue was specifically about investigation, I don't know how (or even if) this is a problem for k8s. I assume increased memory isn't great, but I don't know how that trades against decreased latency for you. If it's the wrong trade, then I would recommend experimenting with GOGC for apiserver. To get back to the old memory footprint, it should work to set GOGC to roughly

(33000 * (1 + oldGOGC/100) / 40425 - 1) * 100
If you're currently using the default GOGC of 100, that would mean setting GOGC to 63.
```

![image](https://user-images.githubusercontent.com/10743879/156120775-a5e8ead9-9007-4709-81a4-5e7c63253b62.png)



So the next step on our side is to run those kind of experiments just to have full confirmation.
Once done, we need two decisions:
- do we want to adjust GOGC for large clusters (I think that way better latencies that we get with this change suggest that maybe not)
- do we want to adjust GOGC for all clusters (to ensure that the cost of running control-plane is not increased to much)

**Comment by wojtek-t** [comment]
We've run a couple tests with GOGC=63, 75 and 90 to confirm all the above hypothesis and the finding is that:
- GOGC=63 - brings back memory usage to the original point (for our particular 5k test)
- GOGC=75,90 - bring it down but by smaller amounts respectively
- even with GOGC=63, we observe visible drop in API call latencies (not as huge as with default GOGC=100, but still visible)


Thinking more about this, I think we probably don't want to take any specific actions out-of-the-box. My personal recommendation would be to just add an `action-required` release note, somewhat along the lines below

```release-note
Kubernetes 1.24 bumped version of golang it is compiled with to go1.18, which introduced significant changes to its garbage collection algorithm. As a result, we observed an increase in memory usage for kube-apiserver in larger an heavily loaded clusters up to ~25% (with the benefit of huge drop in API call latencies). If the memory increase is not acceptable for you you can mitigate by setting GOGC env variable (for our tests using GOGC=63 brings memory usage back to original value, although the exact value may depend on usage patterns on your cluster).
```

[I would appreciate someone adjusting the working probably.]

@liggitt @justaugustus - WDYT ^^

**Comment by liggitt** [comment]
I'd recommend being more specific in the release note (quantify the latency drop we saw instead of `huge`), but yeah... surfacing the default time/space tradeoff has changed in golang and you may need to use GOGC to tune that seems reasonable

**Comment by dims** [comment]
@liggitt @wojtek-t looks a bunch of projects use `debug.SetGCPercent` to customize `GOGC` example [1]. Do we want to add a command line param to expose this explicitly? (so then we can document that folks can tweak that command line param)

[1] https://github.com/containerd/containerd/search?q=SetGCPercent

**Comment by liggitt** [comment]
my first reaction is that I would rather not get in the business of plumbing options to the runtime that can already be set via envvars

**Comment by dims** [comment]
gotcha. at least we should log it so will be easier when folks open issues, we can ask them to grep for it.

**Comment by liggitt** [comment]
That seems like a good idea in general (possibly other things like GOMAXPROCS, too, etc)

**Comment by wojtek-t** [comment]
> I'd recommend being more specific in the release note (quantify the latency drop we saw instead of huge), but yeah... surfacing the default time/space tradeoff has changed in golang and you may need to use GOGC to tune that seems reasonable.

Thanks - so modifying my proposal for release note:
```
Kubernetes 1.24 bumped version of golang it is compiled with to go1.18, which introduced significant changes to its garbage collection algorithm. As a result, we observed an increase in memory usage for kube-apiserver in larger an heavily loaded clusters up to ~25% (with the benefit of API call latencies drop by up to 10x on 99th percentiles). If the memory increase is not acceptable for you you can mitigate by setting GOGC env variable (for our tests using GOGC=63 brings memory usage back to original value, although the exact value may depend on usage patterns on your cluster).
```

Reference to the above 10x:
https://perf-dash.k8s.io/#/?jobname=gce-5000Nodes&metriccategoryname=APIServer&metricname=LoadResponsiveness_PrometheusSimple&Resource=nodes&Scope=resource&Subresource=&Verb=GET

> gotcha. at least we should log it so will be easier when folks open issues, we can ask them to grep for it.

I opened https://github.com/kubernetes/kubernetes/pull/108480 for it - if we agree that's the right way, will add to other components too.

**Comment by jyotimahapatra** [comment]
@justaugustus I'm from the bug triage team. I wanted to check if there are more tasks needed to track this to closure. Looks like https://github.com/kubernetes/kubernetes/pull/108480 is merged already.

**Comment by wojtek-t** [comment]
> @justaugustus I'm from the bug triage team. I wanted to check if there are more tasks needed to track this to closure. Looks like https://github.com/kubernetes/kubernetes/pull/108480 is merged already.

Yes - we want the release note of action-required type that I mentioned above. But we should probably add that on the PR that will be bumping the golang to 1.18 (from the current 1.18-rc)  - although I don't have strong opinion here.

**Comment by wojtek-t** [comment]
I have updated the release note in https://github.com/kubernetes/kubernetes/pull/108870 - resolving.

**Comment by xigang** [comment]
Follow this issue.
